java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp

```diff
@@ -51,6 +51,56 @@ from the input stream removes the vulnerability.
 
 </example>
 
+<p>
+
+Fixes by framework
+<table>
+    <tbody>
+        <tr>
+            <th>Project</th>
+            <th>Maven Cordinates</th>
+            <th>Secure by Default</th>
+            <th>Fix</th>
+        </tr>
+            <tr>
+            <td>XMLDecoder</td>
+            <td>Java Standard Library</td>
+            <td>No</td>
+            <td>Don't use XMLDecoder with untrusted user input. It is impossible to secure.</td>
+        </tr>
+        <tr>
+            <td>ObjectInputStream</td>
+            <td>Java Standard Library</td>
+            <td>No</td>
+            <td>Leverage a validating input stream like <code>org.apache.commons.io.serialization.ValidatingObjectInputStream</code></td>
+        </tr>
+            <tr>
+            <td>FastJson</td>
+            <td>com.alibaba:fastjson</td>
+            <td>Partially</td>
+            <td>Call <code>com.alibaba.fastjson.parser.ParserConfig#setSafeMode</code> with the argument <code>true</code></td>
+        </tr>
+        <tr>
+            <td>SnakeYAML</td>
+            <td>org.yaml:snakeyaml</td>
+            <td><a href="https://bitbucket.org/snakeyaml/snakeyaml/wiki/CVE%20&amp;%20NIST.md">No</a>. <a href="https://bitbucket.org/snakeyaml/snakeyaml/issues/561/cve-2022-1471-vulnerability-in">Maintainer response</a>.</td>
+            <td>Instantiate the <code>org.yaml.snakeyaml.Yaml</code> instance explicitly with an instance of <code>org.yaml.snakeyaml.constructor.SafeConstructor</code> as an argument.</td>
+        </tr>
+        <tr>
+            <td>FasterXML jackson-databind</td>
+            <td>com.fasterxml.jackson.core:jackson-databind</td>
+            <td>Yes</td>
+            <td>Don't call <code>com.fasterxml.jackson.databind.ObjectMapper#enableDefaultTyping</code>. Read <a href="https://cowtowncoder.medium.com/jackson-2-10-safe-default-typing-2d018f0ce2ba">this guide</a>.</td>
+        </tr>
+        <tr>
+            <td>Kryo</td>
+            <td>com.esotericsoftware:kryo and com.esotericsoftware:kryo5</td>
+            <td>com.esotericsoftware:kryo versions after 5.0 Yes; com.esotericsoftware:kryo5 Yes</td>
+            <td>Don't call <code>com.esotericsoftware.kryo(5).Kryo#setRegistrationRequired</code> with the argument <code>false</code>.</td>
+        </tr>
```

Comment:
I did some research on this and Kryo is no longer vulnerable by default. It seems that sometime between the end of the 4.x release line and the latest release this was fixed to be secure by default.

Currently, the Kryo query still present in CodeQL checks for this incorrectly.

CodeQL currently attempts to detect calls to `setRegistrationRequired(true)` for the use case to be safe. But actually, the library currently requires the user to call `setRegistrationRequired(false)` to be vulnerable. The logic in CodeQL needs to be inverted to be correct.

Reply:
I'm trying to track down the exact version this was fixed. I've reached out to the maintainers here: https://github.com/EsotericSoftware/kryo/issues/929

Reply:
Confirmed that versions 5.0.0 and later are all secure-by-default.